### PR TITLE
Migrations API refactoring

### PR DESCRIPTION
**Pull request by radex**
- Refactored migration spec API based on feedback from @mobily:
  - `minimumVersion`, `currentVersion` parameters are deleted from `schemaMigrations()`
  - `{ from: 3, to: 4, steps: [] }` is simplified to `{ toVersion: 4, steps: [] }`. Skipping multiple versions is not allowed
  - @mobily suggested `version: x`, but I decided that `toVersion: x` is more obvious, since this isn't version 4 _of the migration_, but a migration of the schema from version 3 _to version 4_.
- Updated tests & refactored code based on those changes

### CODE REVIEW

**Comment by radex** [comment]
@kokusGr CR please ;)

File: src/Schema/migrations/test.js [hunk 1]
```diff
@@ -78,94 +97,51 @@ describe('schemaMigrations()', () => {
...
-        currentVersion: 3,
-        migrations: [{ from: 1, to: 2, steps: [] }],
+        migrations: [
+          { toVersion: 5, steps: [] },
+          { toVersion: 4, steps: [] },
+          { toVersion: 2, steps: [] },
+        ],
       }),
-    ).toThrow(/cover schema versions/)
+    ).toThrow(/in reverse chronological order/)
```
Comment (kokusGr) [inline src/Schema/migrations/test.js]: Shouldn't this throw something like `gap` or `missing step`?

File: src/Schema/migrations/index.js [hunk 2]
```diff
@@ -83,51 +81,54 @@ export type SchemaMigrations = $Exact<{
...
+      }
+      maxCoveredVersion = toVersion
     })
-    invariant(
-      maxCoveredRange === currentVersion,
-      `Invalid migrations! schemaMigrations() says the current version is ${currentVersion}, but migrations listed only cover schema versions range from ${minimumVersion} to ${maxCoveredRange}. Remember that migrations must be listed in reverse chronological order`,
-    )
   }
+
+  const sortedMigrations = sortMigrations(migrations)
```
Comment (kokusGr) [inline src/Schema/migrations/index.js]: Why migrations have to be in a reverse chronological order if we are sorting it anyway? We can first sort it and then check for completeness.

**Comment by kokusGr** [comment]
@radex Looks good, only few questions :)

File: src/Schema/migrations/index.js [hunk 3]
```diff
@@ -83,51 +81,54 @@ export type SchemaMigrations = $Exact<{
...
+  const oldestMigration = head(sortedMigrations)
+  const newestMigration = last(sortedMigrations)
+  const minVersion = oldestMigration ? oldestMigration.toVersion - 1 : 1
+  const maxVersion = newestMigration ? newestMigration.toVersion : 1
```
[the remark this answers is not in the document]
Comment (radex) [reply src/Schema/migrations/index.js]: Oh I thought we don't support `?.` in Watermelon but I see it's already used in Relation

File: src/Schema/migrations/index.js (hunk 2 above)
Comment (radex) [reply src/Schema/migrations/index.js]: yeah, artifact of previous implementation. If you think this is going to be annoying, I can change it.

Comment (kokusGr) [reply src/Schema/migrations/index.js]: I don't think so. Most people will do it anyway :) But maybe we don't have to force reverse order? Some users might find chronological order better.

Comment (radex) [reply src/Schema/migrations/index.js]: OK, fixed

**Comment by radex** [comment]
@kokusGr changed as to not force a particular order

File: src/Schema/migrations/index.js [hunk 4]
```diff
@@ -83,51 +78,56 @@ export type SchemaMigrations = $Exact<{
...
+      const { toVersion } = migration
+      if (maxCoveredVersion) {
+        invariant(
+          toVersion === maxCoveredVersion + 1,
+          `Invalid migrations! Migrations listed cover range from version ${minVersion} to ${maxCoveredVersion}, but migration ${JSON.stringify(
+            migration,
+          )} is to version ${toVersion}. Migrations must be listed without gaps, or duplicates.`,
+        )
+      }
+      maxCoveredVersion = toVersion
```
Comment (mobily) [inline src/Schema/migrations/index.js]: I guess we don't need the `maxCoveredVersion` variable, use the `reduce` function instead of `forEach`, please.

Comment (radex) [reply src/Schema/migrations/index.js]: Oh, Marcin, you… never miss an opportunity to inject a `reduce` 😂

**Comment by radex** [comment]
@mobily 🏓

**Comment by mobily** [comment]
thanks @radex! 🎉

**Pull request merged**